File: src/python/xerxes/tools/google_search.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Google search integration supporting both API and web scraping backends.

This module provides Google search functionality through the Custom Search API
(preferred when credentials are available) or HTML scraping fallback.

Example:
    >>> from xerxes.tools.google_search import GoogleSearch, configure_google_search
    >>> configure_google_search(api_key="...", cse_id="...")
    >>> results = GoogleSearch.static_call(query="machine learning")
"""

from __future__ import annotations

import logging
import os
import re
import typing as tp
import urllib.parse
from dataclasses import dataclass

from ..types import AgentBaseFn

logger = logging.getLogger(__name__)
SCRAPE_USER_AGENT = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15"
)


@dataclass
class GoogleSearchConfig:
    """Configuration for Google search operations.

    Attributes:
        api_key: Google API key for Custom Search API.
        cse_id: Custom Search Engine ID.
        api_base: Base URL for the API endpoint.
        scrape_base: Base URL for scraping.
        safe: Safe search setting ('active' or 'off').
        user_agent: User agent string for scraping.
    """

    api_key: str = ""
    cse_id: str = ""
    api_base: str = "https://www.googleapis.com/customsearch/v1"
    scrape_base: str = "https://www.google.com/search"
    safe: str = "off"
    user_agent: str = SCRAPE_USER_AGENT


_config = GoogleSearchConfig(
    api_key=os.environ.get("GOOGLE_API_KEY", ""),
    cse_id=os.environ.get("GOOGLE_CSE_ID", ""),
)
_http_client: tp.Any | None = None


def configure_google_search(
    *,
    api_key: str | None = None,
    cse_id: str | None = None,
    safe: str | None = None,
    user_agent: str | None = None,
) -> GoogleSearchConfig:
    """Configure Google search settings.

    Args:
        api_key: Google API key. Overrides environment variable.
        cse_id: Custom Search Engine ID. Overrides environment variable.
        safe: Safe search setting ('active' or 'off').
        user_agent: Custom user agent string.

    Returns:
        Updated GoogleSearchConfig.

    Example:
        >>> configure_google_search(
        ...     api_key="your-api-key",
        ...     cse_id="your-cse-id"
        ... )
    """
    global _config
    if api_key is not None:
        _config.api_key = api_key
    if cse_id is not None:
        _config.cse_id = cse_id
    if safe is not None:
        _config.safe = safe
    if user_agent is not None:
        _config.user_agent = user_agent
    return _config


def get_google_search_config() -> GoogleSearchConfig:
    """Get the current Google search configuration.

    Returns:
        The current GoogleSearchConfig instance.
    """
    return _config


def set_google_search_client(client: tp.Any | None) -> None:
    """Set a custom HTTP client for Google searches.

    Args:
        client: Custom HTTP client instance, or None to use default httpx.
    """
    global _http_client
    _http_client = client


def _http_get(url: str, *, headers: dict[str, str] | None = None, params: dict[str, str] | None = None) -> tp.Any:
    """Perform HTTP GET request using configured client.

    Args:
        url: URL to fetch.
        headers: Optional HTTP headers.
        params: Optional query parameters.

    Returns:
        HTTP response object.
    """
    if _http_client is not None:
        return _http_client.get(url, headers=headers, params=params)
    try:
        import httpx
    except ImportError as exc:
        raise RuntimeError("httpx required for GoogleSearch HTTP fallback") from exc
    return httpx.get(url, headers=headers or {}, params=params or {}, timeout=20.0, follow_redirects=True)


def _resp_text(resp: tp.Any) -> str:
    """Extract text content from HTTP response.

    Args:
        resp: HTTP response object.

    Returns:
        Response body as string.
    """
    text = getattr(resp, "text", None)
    if isinstance(text, str):
        return text
    body = getattr(resp, "body", "") or ""
    if isinstance(body, bytes):
        body = body.decode(errors="replace")
    return body


def _resp_json(resp: tp.Any) -> dict[str, tp.Any]:
    """Parse JSON from HTTP response.

    Args:
        resp: HTTP response object.

    Returns:
        Parsed JSON as dictionary.
    """
    if hasattr(resp, "json") and callable(resp.json):
        try:
            return resp.json()
        except Exception:
            pass
    text = _resp_text(resp)
    try:
        import json

        return json.loads(text)
    except Exception:
        return {}


def _resp_status(resp: tp.Any) -> int:
    """Get HTTP status code from response.

    Args:
        resp: HTTP response object.

    Returns:
        Status code as integer.
    """
    return int(getattr(resp, "status_code", 0) or 0)


def _search_via_api(
    query: str,
    *,
    n_results: int,
    cfg: GoogleSearchConfig,
    site: str | None,
    time_range: str | None,
) -> dict[str, tp.Any]:
    """Search using Google Custom Search API.

    Args:
        query: Search query string.
        n_results: Number of results to return.
        cfg: Search configuration.
        site: Optional site restriction.
        time_range: Optional time range filter.

    Returns:
        Dictionary with search results and metadata.
    """
    params: dict[str, str] = {
        "key": cfg.api_key,
        "cx": cfg.cse_id,
        "q": (f"site:{site} " if site else "") + query,
        "num": str(min(max(n_results, 1), 10)),
        "safe": "active" if cfg.safe == "active" else "off",
    }
    if time_range:
        params["dateRestrict"] = time_range
    resp = _http_get(cfg.api_base, params=params)
    status = _resp_status(resp)
    if status and status >= 400:
        return {
            "engine": "google_api",
            "query": query,
            "error": f"HTTP {status}",
            "results": [],
        }
    payload = _resp_json(resp)
    items = payload.get("items") or []
    out_items: list[dict[str, tp.Any]] = []
    for it in items[:n_results]:
        out_items.append(
            {
                "title": it.get("title", ""),
                "url": it.get("link", ""),
                "snippet": it.get("snippet", ""),
                "displayed_url": it.get("displayLink", ""),
            }
        )
    return {
        "engine": "google_api",
        "query": query,
        "count": len(out_items),
        "results": out_items,
        "search_information": payload.get("searchInformation", {}),
    }


_RESULT_RE = re.compile(
    r'<a[^>]+href="(?P<url>https?://[^"#&]+)"[^>]*>'
    r".*?<h3[^>]*>(?P<title>.*?)</h3>"
    r'(?:.*?<div[^>]*class="VwiC3b[^"]*"[^>]*>(?P<snippet>.*?)</div>)?',
    re.DOTALL,
)


def _strip_tags(html: str) -> str:
    """Remove HTML tags from string.

    Args:
        html: HTML string to strip.

    Returns:
        Plain text with HTML tags removed.
    """
    return re.sub(r"<[^>]+>", "", html or "").strip()


def _search_via_scrape(
    query: str,
    *,
    n_results: int,
    cfg: GoogleSearchConfig,
    site: str | None,
    time_range: str | None,
) -> dict[str, tp.Any]:
    """Search by scraping Google HTML.

    Args:
        query: Search query string.
        n_results: Number of results to return.
        cfg: Search configuration.
        site: Optional site restriction.
        time_range: Optional time range filter.

    Returns:
        Dictionary with search results and metadata.
    """
    q = (f"site:{site} " if site else "") + query
    params: dict[str, str] = {
        "q": q,
        "num": str(min(max(n_results, 1), 30)),
        "hl": "en",
    }
    if cfg.safe in ("active", "on"):
        params["safe"] = "active"
    if time_range:
        params["tbs"] = f"qdr:{time_range[0]}"
    headers = {
        "User-Agent": cfg.user_agent,
        "Accept-Language": "en-US,en;q=0.9",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    }
    try:
        resp = _http_get(cfg.scrape_base, headers=headers, params=params)
    except Exception as exc:
        return {"engine": "google_scrape", "query": query, "error": str(exc), "results": []}
    status = _resp_status(resp)
    html = _resp_text(resp)
    if status and status >= 400:
        return {
            "engine": "google_scrape",
            "query": query,
            "error": f"HTTP {status} (Google likely blocked the scrape — set GOOGLE_API_KEY+GOOGLE_CSE_ID)",
            "results": [],
        }
    out: list[dict[str, str]] = []
    _BeautifulSoup: type | None = None
    try:
        from bs4 import BeautifulSoup, Tag

        _BeautifulSoup = BeautifulSoup
    except ImportError:
        pass
    if _BeautifulSoup is not None:
        soup = _BeautifulSoup(html, "html.parser")
        for h3 in soup.find_all("h3")[: n_results * 4]:
            link = h3.find_parent("a")
            if not link:
                continue
            tag_link = tp.cast(Tag, link)
            url = str(tag_link.get("href", ""))
            if not url.startswith(("http://", "https://")):
                continue
            host = urllib.parse.urlparse(url).netloc.lower()
            if host.endswith(("google.com", "googleusercontent.com", "gstatic.com")) or host.startswith("webcache."):
                continue
            title = h3.get_text(" ", strip=True)
            snippet = ""
            container = tag_link.find_parent(class_=lambda c: bool(c and "g" in c.split()))
            if container is None:
                container = tag_link
            tag_container = tp.cast(Tag, container)
            for cand in tag_container.find_all(class_=lambda c: bool(c and ("VwiC3b" in c or "MUxGbd" in c))):
                txt = cand.get_text(" ", strip=True)
                if txt and len(txt) > 20:
                    snippet = txt[:300]
                    break
            out.append({"title": title, "url": url, "snippet": snippet})
            if len(out) >= n_results:
                break
    if not out:
        for m in _RESULT_RE.finditer(html):
            url = m.group("url")
            if "google.com" in url or url.startswith("https://webcache"):
                continue
            title = _strip_tags(m.group("title"))
            snippet = _strip_tags(m.group("snippet") or "")
            out.append({"title": title, "url": url, "snippet": snippet[:300]})
            if len(out) >= n_results:
                break
    return {
        "engine": "google_scrape",
        "query": query,
        "count": len(out),
        "results": out,
        "warning": (
            "Anonymous scrapes are rate-limited; set GOOGLE_API_KEY + GOOGLE_CSE_ID for a quota-backed path."
            if not out
            else ""
        ),
    }


class GoogleSearch(AgentBaseFn):
    """Perform Google searches using API or scraping.

    Uses Google Custom Search API when credentials are configured,
    falls back to HTML scraping otherwise.

    Example:
        >>> results = GoogleSearch.static_call(
        ...     query="Python tutorials",
        ...     n_results=10
        ... )
    """

    @staticmethod
    def static_call(
        query: str,
        n_results: int = 5,
        site: str | None = None,
        time_range: str | None = None,
        **context_variables: tp.Any,
    ) -> dict[str, tp.Any]:
        """Perform a Google search.

        Args:
            query: Search query string.
            n_results: Number of results to return. Defaults to 5.
            site: Optional site to restrict search to.
            time_range: Optional time filter ('d' for day, 'w' for week, 'm' for month, 'y' for year).
            **context_variables: Additional context passed through to downstream calls.

        Returns:
            Dictionary with search results containing 'results' list and metadata.

        Example:
            >>> GoogleSearch.static_call("AI research", n_results=10)
        """
        if isinstance(n_results, str):
            try:
                n_results = int(n_results)
            except ValueError:
                n_results = 5
        return GoogleSearch._execute(query, n_results, site, time_range)

    @staticmethod
    def _execute(
        query: str,
        n_results: int = 5,
        site: str | None = None,
        time_range: str | None = None,
    ) -> dict[str, tp.Any]:
        """Internal search execution with backend selection.

        Args:
            query: Search query.
            n_results: Number of results.
            site: Optional site filter.
            time_range: Optional time filter.

        Returns:
            Search results dictionary.
        """
        cfg = get_google_search_config()
        if cfg.api_key and cfg.cse_id:
            return _search_via_api(query, n_results=n_results, cfg=cfg, site=site, time_range=time_range)
        return _search_via_scrape(query, n_results=n_results, cfg=cfg, site=site, time_range=time_range)


__all__ = [
    "GoogleSearch",
    "GoogleSearchConfig",
    "configure_google_search",
    "get_google_search_config",
    "set_google_search_client",
]

File: xerxes/src/tools/googleSearch.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { ConfigurationError, ValidationError } from '../core/errors.js'
import { ToolRegistry } from '../executors/toolRegistry.js'
import type { JsonObject, JsonValue, ToolDefinition } from '../types/toolCalls.js'
import { optionalInteger, optionalString, requiredString } from './inputs.js'
import { PublicWebClient } from './webTools.js'

const GOOGLE_CUSTOM_SEARCH_ENDPOINT = 'https://www.googleapis.com/customsearch/v1'
const GOOGLE_RESULT_LIMIT = 10

export interface GoogleCustomSearchOptions {
  readonly apiKey: string
  readonly endpoint?: string
  readonly safeSearch?: 'active' | 'off'
  readonly searchEngineId: string
  readonly webClient?: PublicWebClient
}

export interface GoogleSearchRequest {
  readonly nResults?: number
  readonly query: string
  readonly site?: string
  readonly timeRange?: string
}

export interface GoogleSearchResult {
  readonly displayedUrl: string
  readonly snippet: string
  readonly title: string
  readonly url: string
}

export interface GoogleSearchResponse {
  readonly count?: number
  readonly engine: 'google_api'
  readonly error?: string
  readonly query: string
  readonly results: readonly GoogleSearchResult[]
  readonly searchInformation?: JsonObject
}

/** Google Custom Search JSON API client. It intentionally has no anonymous HTML-scraping fallback. */
export class GoogleCustomSearchClient {
  private readonly apiKey: string
  private readonly endpoint: string
  private readonly safeSearch: 'active' | 'off'
  private readonly searchEngineId: string
  private readonly webClient: PublicWebClient

  constructor(options: GoogleCustomSearchOptions) {
    this.apiKey = requiredConfiguration(options.apiKey, 'apiKey')
    this.searchEngineId = requiredConfiguration(options.searchEngineId, 'searchEngineId')
    this.endpoint = options.endpoint ?? GOOGLE_CUSTOM_SEARCH_ENDPOINT
    this.safeSearch = options.safeSearch ?? 'off'
    this.webClient = options.webClient ?? new PublicWebClient()
  }

  async search(request: GoogleSearchRequest, signal?: AbortSignal): Promise<GoogleSearchResponse> {
    const query = request.query.trim()
    if (!query) throw new ValidationError('query', 'must be non-empty', request.query)
    const nResults = requestedResultCount(request.nResults ?? 5)
    const url = this.requestUrl(query, nResults, request.site, request.timeRange)
    const fetched = await this.webClient.fetch(url, { method: 'GET' }, signal === undefined ? {} : { signal })
    if (!fetched.response.ok) {
      return { engine: 'google_api', error: `HTTP ${fetched.response.status}`, query, results: [] }
    }

    const responseText = await this.webClient.text(fetched.response)
    let payload: JsonObject
    try {
      payload = parseObject(responseText)
    } catch {
      return { engine: 'google_api', error: 'Google Custom Search returned invalid JSON', query, results: [] }
    }
    const results = googleResults(payload.items, nResults)
    const response: {
      count: number
      engine: 'google_api'
      query: string
      results: readonly GoogleSearchResult[]
      searchInformation?: JsonObject
    } = {
      count: results.length,
      engine: 'google_api',
      query,
      results,
    }
    const searchInformation = payload.searchInformation
    if (isObject(searchInformation)) response.searchInformation = searchInformation
    return response
  }

  private requestUrl(query: string, nResults: number, site: string | undefined, timeRange: string | undefined): string {
    let endpoint: URL
    try {
      endpoint = new URL(this.endpoint)
    } catch {
      throw new ConfigurationError('endpoint', 'must be an absolute Google Custom Search API URL')
    }
    if (endpoint.protocol !== 'http:' && endpoint.protocol !== 'https:') {
      throw new ConfigurationError('endpoint', 'must use HTTP or HTTPS')
    }
    const sitePrefix = site?.trim()
    endpoint.searchParams.set('key', this.apiKey)
    endpoint.searchParams.set('cx', this.searchEngineId)
    endpoint.searchParams.set('q', sitePrefix ? `site:${sitePrefix} ${query}` : query)
    endpoint.searchParams.set('num', String(nResults))
    endpoint.searchParams.set('safe', this.safeSearch)
    if (timeRange !== undefined) {
      if (!/^[dwmy]\d*$/i.test(timeRange)) {
        throw new ValidationError('time_range', 'must use Google dateRestrict syntax such as d, w, m, y, or m6', timeRange)
      }
      endpoint.searchParams.set('dateRestrict', timeRange)
    }
    return endpoint.toString()
  }
}

export const GOOGLE_SEARCH_DEFINITION: ToolDefinition = {
  type: 'function',
  function: {
    name: 'GoogleSearch',
    description: 'Search through a configured Google Custom Search JSON API key and search-engine id.',
    parameters: {
      type: 'object',
      additionalProperties: false,
      properties: {
        query: { type: 'string' },
        n_results: { type: 'integer', minimum: 1, maximum: GOOGLE_RESULT_LIMIT, default: 5 },
        site: { type: 'string' },
        time_range: { type: 'string', description: 'Google dateRestrict value such as d, w, m, y, or m6.' },
      },
      required: ['query'],
    },
  },
}

/** Register GoogleSearch only after the host explicitly provides API credentials. */
export function registerGoogleSearchTool(
  registry: ToolRegistry,
  search: GoogleCustomSearchClient,
  agentId = 'default',
): void {
  registry.register(GOOGLE_SEARCH_DEFINITION, (inputs, _context, signal) => search.search(googleRequestFromInputs(inputs), signal), agentId)
}

export function googleSearchLimitations(): readonly string[] {
  return Object.freeze([
    'GoogleSearch requires an explicit Google Custom Search API key and search-engine id.',
    'Anonymous Google HTML scraping is deliberately not implemented.',
  ])
}

function requiredConfiguration(value: string, name: string): string {
  const normalized = value.trim()
  if (!normalized) throw new ConfigurationError(name, 'must be explicitly configured')
  return normalized
}

function requestedResultCount(value: number): number {
  if (!Number.isInteger(value) || value < 1) {
    throw new ValidationError('n_results', 'must be a positive integer', value)
  }
  return Math.min(value, GOOGLE_RESULT_LIMIT)
}

function googleRequestFromInputs(inputs: JsonObject): GoogleSearchRequest {
  const request: { nResults: number; query: string; site?: string; timeRange?: string } = {
    nResults: optionalInteger(inputs, 'n_results', 5),
    query: requiredString(inputs, 'query'),
  }
  const site = optionalString(inputs, 'site')
  const timeRange = optionalString(inputs, 'time_range')
  if (site !== undefined) request.site = site
  if (timeRange !== undefined) request.timeRange = timeRange
  return request
}

function googleResults(value: JsonValue | undefined, maximum: number): GoogleSearchResult[] {
  if (!Array.isArray(value)) return []
  const results: GoogleSearchResult[] = []
  for (const item of value) {
    if (!isObject(item)) continue
    const url = stringValue(item.link)
    if (!url) continue
    results.push({
      displayedUrl: stringValue(item.displayLink),
      snippet: stringValue(item.snippet),
      title: stringValue(item.title),
      url,
    })
    if (results.length >= maximum) break
  }
  return results
}

function parseObject(value: string): JsonObject {
  const parsed = JSON.parse(value) as JsonValue
  if (!isObject(parsed)) throw new Error('expected object')
  return parsed
}

function isObject(value: JsonValue | undefined): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function stringValue(value: JsonValue | undefined): string {
  return typeof value === 'string' ? value : ''
}
